src/legacy/ui/ui_settings/ui_settings_service.ts

```diff
@@ -16,28 +16,77 @@
  * specific language governing permissions and limitations
  * under the License.
  */
-
+import { Legacy } from 'kibana';
 import { defaultsDeep } from 'lodash';
 import Boom from 'boom';
 
+import { SavedObjectsClientContract, SavedObjectAttribute } from 'src/core/server';
 import { createOrUpgradeSavedConfig } from './create_or_upgrade_saved_config';
 
+export interface UiSettingsServiceOptions {
+  type: string;
+  id: string;
+  buildNum: number;
+  overrides: Record<string, SavedObjectAttribute>;
+  savedObjectsClient: SavedObjectsClientContract;
+  getDefaults?: () => Record<string, UiSettingsParams>;
+  logWithMetadata?: Legacy.Server['logWithMetadata'];
+}
+
+interface ReadOptions {
+  ignore401Errors?: boolean;
+  autoCreateOrUpgradeIfMissing?: boolean;
+}
+
+interface UserProvidedValue {
+  userValue?: SavedObjectAttribute;
+  isOverridden?: boolean;
+}
+
+type UiSettingsRawValue = UiSettingsParams & UserProvidedValue;
+
+type UserProvided = Record<string, UserProvidedValue>;
+type UiSettingsRaw = Record<string, UiSettingsRawValue>;
+
+type UiSettingsType = 'json' | 'markdown' | 'number' | 'select' | 'boolean' | 'string';
+
+interface UiSettingsParams {
+  name: string;
+  value: SavedObjectAttribute;
+  description: string;
+  category: string[];
+  options?: string[];
+  optionLabels?: Record<string, string>;
+  requiresPageReload?: boolean;
+  readonly?: boolean;
+  type?: UiSettingsType;
+}
+
+export interface IUiSettingsService {
+  getDefaults: () => Promise<Record<string, UiSettingsParams>>;
+  get: <T extends SavedObjectAttribute = any>(key: string) => Promise<T>;
+  getAll: <T extends SavedObjectAttribute = any>() => Promise<Record<string, T>>;
+  getUserProvided: () => Promise<UserProvided>;
+  setMany: <T extends SavedObjectAttribute = any>(changes: Record<string, T>) => Promise<void>;
+  set: <T extends SavedObjectAttribute = any>(key: string, value: T) => Promise<void>;
+  remove: (key: string) => Promise<void>;
+  removeMany: (keys: string[]) => Promise<void>;
+  isOverridden: (key: string) => boolean;
+}
 /**
  *  Service that provides access to the UiSettings stored in elasticsearch.
  *  @class UiSettingsService
  */
-export class UiSettingsService {
-  /**
-   *  @constructor
-   *  @param {Object} options
-   *  @property {string} options.type type of SavedConfig object
-   *  @property {string} options.id id of SavedConfig object
-   *  @property {number} options.buildNum
-   *  @property {SavedObjectsClient} options.savedObjectsClient
-   *  @property {Function} [options.getDefaults]
-   *  @property {Function} [options.log]
-   */
-  constructor(options) {
+export class UiSettingsService implements IUiSettingsService {
```

Comment:
Shall we rename it to `UiSettingsClient`? It's not a service but a designated part for the plugins to interact with.

Reply:
Yeah I think that makes sense, or even just `UISettings`?

Reply:
Actually, we probably need to decide on our convention for capitalization of acronyms. For example we use `HttpServiceSetup` instead of `HTTPServiceSetup` in core right now. Should this be `UiSettingsClient`? I prefer this pattern because then the I-prefix looks a little less strange: `IUiSettingsClient`